Sharding a Conversation Tutorial (#10)
talkjs/tutorials: how-to-shard-a-conversation/how-to-shard-a-group-chat-in-talkjs.md

# How to shard a group chat in TalkJS

> Sharding is conventionally used in distributed systems to spread users across services to ensure consistent availability. Sharding involves distributing users between available services so that no one service experiences unnecessary loads at any given time. Using TalkJS, you can have group chats with up to 1250 users, [depending on your plan](https://talkjs.com/docs/Features/Group_Chats/#user-limits-in-group-chats). But, if you require more users, you can shard a conversation and assign users to the sharded conversations.
> 
> Sharding conversations is useful in scenarios with a large number of participants where it is not important that users respond back to specific other participants. Examples of this are live streaming, social chats, etc. In our demo, we assign a specific user to a specific shard always so that they're able to retrieve their messages across devices. 
> 
> <figure class="kg-image-card">
>   <img class="kg-image" src="1-random-assignment.png" alt="Users are randomly assigned to conversations."/>
>   <figcaption>After entering their information, users are randomly assigned to conversations.</figcaption>
> </figure>
> 
> The following sections highlight the main topics that we'll cover today:

Reviewer: ```suggestion
Sharding is conventionally used in distributed systems to spread users across services to ensure consistent availability. Sharding involves distributing users between available services so that no one service experiences unnecessary loads at any given time. Using TalkJS, you can have group chats with up to 1250 users, [depending on your plan](https://talkjs.com/docs/Features/Group_Chats/#user-limits-in-group-chats). But, if you require more users, you can shard a conversation and assign users to the sharded conversations.

Conversation sharding is useful in scenarios like livestreams and social chats where there are a large number of users adding short messages in the chat and it is not so important that every user can see all messages.

<figure class="kg-image-card">
  <img class="kg-image" src="1-random-assignment.png" alt="Users are randomly assigned to conversations."/>
  <figcaption>After entering their information, users are randomly assigned to conversations.</figcaption>
</figure>

In our demo, we assign each user to a specific shard based on their id, so that they're able to retrieve their messages across devices.  The following sections highlight the main topics that we'll cover in this tutorial:
```

- Adding a modal for the user to enter their details to enter a group chat 
- Creating a TalkJS user based on the details entered
- Using the generated user's id to assign them to a predetermined conversation shard

To follow along, you’ll need:

- A [TalkJS account](https://talkjs.com/dashboard/login). TalkJS provides a ready-to-use chat client for your application. Your account gives you access to TalkJS's free development environment.
- An existing TalkJS project using the [JavaScript Chat SDK](https://talkjs.com/docs/Reference/JavaScript_Chat_SDK/). See our [Getting Started](https://talkjs.com/docs/Getting_Started/) guide for an example of how to set this up.

This project uses a script to add three example conversations with 2 users each. This is not required in a production scenario where you already have users and conversations set up.

> We’ll build up the feature step by step in the following sections. If you would rather see the complete example code, see the [GitHub repo]() for this tutorial.

Reviewer: ```suggestion
We’ll build up the feature step by step in the following sections. If you would rather see the complete example code, see the [GitHub repo](https://github.com/talkjs/talkjs-examples/tree/master/howtos/how-to-shard-a-conversation) for this tutorial.
```

  const lastIdSection = userId.substring(userId.lastIndexOf("-") + 1);

  // Convert that from hexadecimal to decimal number, 88424362858084
  const lastIdNumeric = parseInt(lastIdSection, 16);

  // Assign to one of N shards using modulo, in this case 88424362858084 % 3 = 1
  return (lastIdNumeric % shardCount) + 1;
}
```

Using the last block of the user's ID, we first convert it from hexadecimal to decimal, and then do a modulo operation by the `shardCount` and add `1` to retrieve a value in `[1,2,3]`. This ensures that the same user is always assigned to the same shard even when accessing from different devices.

## Conclusion

<figure class="kg-image-card">
  <img class="kg-image" src="2-demo.gif" alt="Assign the user to a pre-determined conversation shard."/>
  <figcaption>Assigning the user to a pre-determined conversation shard.</figcaption>
</figure>

You now have a working demonstration of how to shard a conversation! To recap, in this tutorial you have:

- Added a modal for the user to enter their details
- Create a TalkJS user based on the information entered
- Assign the user to one of the conversation shards

> For the full example code for this tutorial, see our [GitHub repo]().

Reviewer: ```suggestion
For the full example code for this tutorial, see our [GitHub repo](https://github.com/talkjs/talkjs-examples/tree/master/howtos/how-to-shard-a-conversation).
```
